### DISCUSSION TITLE
Why is the VITS model slower than Glow-TTS + HiFi-GAN?

### DISCUSSION THREAD

**Discussion by fijipants** [coqui-ai/TTS#728] [General Q&A] [answered]
Looking at [the paper](https://arxiv.org/pdf/2106.06103.pdf), VITS should be faster than Glow-TTS + HiFi-GAN:

![image](https://user-images.githubusercontent.com/88913682/129430337-be9dcda6-d1c4-4d52-a13e-a2e36101741a.png)

Even the docs say it's [67x faster than real time](https://tts.readthedocs.io/en/latest/models/vits.html).

Yet when I test it locally, these is the results I get:

```
Model               Vocoder              Real-time factor
ljspeech/glow-tts   ljspeech/hifigan_v2  0.09
ljspeech/vits                            0.35
ljspeech/tacotron2  ljspeech/hifigan_v2  0.35
```

So not only is it slower than glow-tts, it's even as slow as taco2 + hifigan, and nowhere near "67x faster than real time".

Am I doing something wrong?

---

Testing commands:

```
tts --model_name "tts_models/en/ljspeech/glow-tts" --vocoder_name "vocoder_models/en/ljspeech/hifigan_v2" --text "..." --out_path glow.wav
tts --model_name "tts_models/en/ljspeech/vits" --text "..." --out_path vits.wav
tts --model_name "tts_models/en/ljspeech/tacotron2-DDC" --vocoder_name "vocoder_models/en/ljspeech/hifigan_v2" --text "..." --out_path taco2.wav
```

---

My environment:
- i7 4790k
- RTX 3090
- Ubuntu
- Coqui v0.2.0 installed via `pip install tts`

**Comment by erogol** [comment]
can you share the code that you used for benchmarking?

**Reply by fijipants** [reply]
I just used the three commands listed in the discussion, then looked at the "real-time factor" they printed out.

For text I put the first 3 sentences of the Navy Seals copy-pasta (excuse the language):

```
What the fuck did you just fucking say about me, you little bitch? I'll have you know I graduated top of my class in the Navy Seals, and I've been involved in numerous secret raids on Al-Quaeda, and I have over 300 confirmed kills. I am trained in gorilla warfare and I'm the top sniper in the entire US armed forces.
```

**Comment by Sadam1195** [comment]
following discussion...

**Comment by lexkoro** [comment]
I think those metrics are measured inferecing on a GPU. 
Try adding the `--use_cuda true` flag.

**Reply by fijipants** [reply]
Here's what I got with `--use_cuda true` on GPU (RTX 3080):

```
Model               Vocoder              Real-time factor
ljspeech/glow-tts   ljspeech/hifigan_v2  0.08
ljspeech/vits                            0.09
ljspeech/tacotron2  ljspeech/hifigan_v2  0.06
```

**Comment by erogol** [comment]
GlowTTS you used above does not apply `add_blank` whereas VITS does. Meaning that VITS adds a blank token between every character which doubles the sequence length.

**Reply by BillyBobQuebec** [reply]
I think you have a typo there, did you mean to say "GlowTTS you used above **doesn't** apply add_blank whereas VITS does."?
because your current statement is using the word "does" for both GlowTTS and VITS... which implies that there is no difference between GlowTTS and VITS in the context you speak of.

**Reply by erogol** [reply]
thx for the 🚩.. edited now

**Comment by BillyBobQuebec** [comment, marked as answer]
The VITS model (At least as described in the paper) uses HifiganV1 which is significantly slower than V2, but offers the highest quality:

![image](https://user-images.githubusercontent.com/74849975/129982196-9ab16ffb-6aed-46e3-bc75-324773d31406.png)

I'm fairly sure that in the VITS paper, they are comparing VITS to GlowTTS+HifiganV1. In that paper's comparison, VITS has a real-time factor that is roughly 2.5 times the speed of GlowTTS+HifiganV1. 

Considering this information, I think it's very reasonable to assume that Glow-TTS in coqui would be much closer in speed, or faster than VITS (since Coqui's Glow-TTS is using the much faster HifiganV2 instead of V1). Especially since Glow-TTS itself only takes 40-60ms to process (according to glow-tts paper), the rest of inference time is dependent on Hifigan.

This seems to be the reason, or at least one of the largest factors, that is causing coqui's glowtts+hifiganV2 to be faster than VITS.

**Reply by fijipants** [reply]
So if I'm understanding correctly, coqui VITS is slower than coqui Glow-TTS for these reasons:
- Coqui VITS adds blank tokens between every character, doubling inference time.
- Coqui Glow-TTS uses Hifigan V2 but the paper uses Hifigan V1 (much slower).

Is that right?

**Reply by BillyBobQuebec** [reply]
I would agree in this being a reason: 
"Coqui Glow-TTS uses Hifigan V2 but the paper uses Hifigan V1 (much slower)." 

But to clarify something about the blank tokens, it seems like the paper also tests inference speed with VITS using double the sequence length of the glow model they are comparing it against, just like Coqui does.

So you would technically be right in saying it's "a reason why coqui VITS is slower than coqui Glow-TTS" but not "a reason why Coqui VITS is slower than the original paper's VITS."

**Reply by fijipants** [reply]
Aren't we still missing some speed?

Hifigan V1 -> Hifigan V2 is about a 4.5x speed increase, so if I remake the original chart with (estimated) Hifigan v1 speeds:

```
Model               Vocoder              Real-time factor
ljspeech/glow-tts   ljspeech/hifigan_v1  0.40
ljspeech/glow-tts   ljspeech/hifigan_v2  0.09
ljspeech/vits                            0.35
```

Coqui VITS is only slightly faster Coqui Glow-TTS + (estimated) Hifigan V1, but the paper says it should be 67/27 ~= 2.5x faster.

Could the remaining speed difference be caused by inferencing on CPU instead of on GPU?

**Reply by BillyBobQuebec** [reply]
Seems like you're mistakenly applying the Hifigan GPU speedup to glow-tts coqui speeds on CPU, and comparing that to the cpu speed for VITS, here would be the corrected way using the Hifigan **CPU** speedup(6.7X), and comparing it to your VITS CPU speed in coqui:

ljspeech/glow-tts   ljspeech/hifigan_v1  0.60
ljspeech/glow-tts   ljspeech/hifigan_v2  0.09
ljspeech/vits                                            0.35

this equates to a 1.7X speedup, not quite the speedup shown in the VITS paper, but comparing this chart to the speedups in the VITS paper would still be flawed since the VITS paper is using a GPU to measure speedups.

Luckily the Hifigan paper includes GPU speed comparison between V1 and V2, and luckily you've also provided gpu benchmarks for coqui, so here is a chart for estimated GPU speeds of Coqui's Glow-TTS+HifiganV1:

ljspeech/glow-tts   ljspeech/hifigan_v1  0.36
ljspeech/glow-tts   ljspeech/hifigan_v2  0.08
ljspeech/vits                                            0.09

Here you can see that VITS is 4X the speed of the estimated GlowTTS+HifiganV1 speed! (when comparing GPU speeds like the paper) This extra speedup is likely due to differences in GPU architecture and sequence lengths during testing, but it's close enough where I think we now confidently know, this is the main reason for the discrepancy between the VITS paper, and your original testing with Coqui.
